traffic_ops/traffic_ops_golang/ats/parentdotconfig.go

```diff
@@ -0,0 +1,1393 @@
+	seen := map[tc.DeliveryServiceName]struct{}{}
+	log.Errorf("DEBUG dupar checking\n")
+	for _, ds := range dses {
+		if _, ok := seen[ds.Name]; ok {
+			log.Errorf("DEBUG dupar %v\n", ds)
+		}
+		seen[ds.Name] = struct{}{}
+	}
+	log.Errorf("DEBUG dupar checked.\n")
+
+	if err != nil {
+		return nil, errors.New("getting raw parent config ds: " + err.Error())
+	}
+
+	dsesWithParams, err := getParentConfigDSParams(tx, dses)
+	if err != nil {
+		return nil, errors.New("getting ds params: " + err.Error())
+	}
+	return dsesWithParams, nil
+}
+
+// my $qstring = $self->profile_param_value( $server_obj->profile->id, 'parent.config', 'qstring', undef );
+// 		my $parent_select_alg = $self->profile_param_value( $server_obj->profile->id, 'parent.config', 'algorithm', undef );
+// my $qsh = $self->profile_param_value( $server_obj->profile->id, 'parent.config', 'psel.qstring_handling');
+
+const ParentConfigParamQStringHandling = "psel.qstring_handling"
+const ParentConfigParamMSOAlgorithm = "mso.algorithm"
+const ParentConfigParamMSOParentRetry = "mso.parent_retry"
+const ParentConfigParamUnavailableServerRetryResponses = "mso.unavailable_server_retry_responses"
+const ParentConfigParamMaxSimpleRetries = "mso.max_simple_retries"
+const ParentConfigParamMaxUnavailableServerRetries = "mso.max_unavailable_server_retries"
+const ParentConfigParamAlgorithm = "algorithm"
+const ParentConfigParamQString = "qstring"
+
+func getParentConfigServerProfileParams(tx *sql.Tx, serverID int) (map[string]string, error) {
+	qry := `
+SELECT
+  pa.name,
+  pa.value
+FROM
+  parameter pa
+  JOIN profile_parameter pp ON pp.parameter = pa.id
+  JOIN profile pr ON pr.id = pp.profile
+  JOIN server s on s.profile = pr.id
+WHERE
+  s.id = $1
+  AND pa.config_file = 'parent.config'
+  AND pa.name IN (
+    '` + ParentConfigParamQStringHandling + `',
+    '` + ParentConfigParamAlgorithm + `',
+    '` + ParentConfigParamQString + `'
+  )
+`
+	rows, err := tx.Query(qry, serverID)
+	if err != nil {
+		return nil, errors.New("querying: " + err.Error())
+	}
+	defer rows.Close()
+	params := map[string]string{}
+	for rows.Next() {
+		name := ""
+		val := ""
+		if err := rows.Scan(&name, &val); err != nil {
+			return nil, errors.New("scanning: " + err.Error())
+		}
+		params[name] = val
+	}
+	return params, nil
+}
+
+func getParentConfigDSRaw(tx *sql.Tx, qry string, qryParams []interface{}) ([]ParentConfigDS, error) {
+	log.Errorf("DEBUG PCGen getParentConfigDSRaw qry qqq%+vqqq params %+v\n", qry, qryParams)
+	rows, err := tx.Query(qry, qryParams...)
+	if err != nil {
+		return nil, errors.New("querying: " + err.Error())
+	}
+	defer rows.Close()
+	dses := []ParentConfigDS{}
+	for rows.Next() {
+		d := ParentConfigDS{}
+		if err := rows.Scan(&d.Name, &d.QStringIgnore, &d.OriginFQDN, &d.MultiSiteOrigin, &d.OriginShield, &d.Type); err != nil {
+			return nil, errors.New("scanning: " + err.Error())
+		}
+		if d.OriginFQDN == "" {
+			// TODO skip ANY_MAP DSes? Why? Did Perl, I didn't see it?
```

Comment:
I've confirmed to the best of my ability. I'd prefer to leave this comment in, because I'm unsure this is the correct behavior, but changing it at this time would be dangerous.